src/multisig/signing_protocol.cpp

```diff
@@ -0,0 +1,967 @@
+      return false;
+  }
+
+  // prepare rings for input CLSAGs
+  rv.mixRing.resize(num_sources);
+  for (std::size_t i = 0; i < num_sources; ++i) {
+    const std::size_t ring_size = sources[i].outputs.size();
+    rv.mixRing[i].resize(ring_size);
+    for (std::size_t j = 0; j < ring_size; ++j) {
+      rv.mixRing[i][j].dest = sources[i].outputs[j].second.dest;
+      rv.mixRing[i][j].mask = sources[i].outputs[j].second.mask;
+    }
+  }
+
+  // make pseudo-output commitments
+  rct::keyV a;  //pseudo-output commitment blinding factors
+  auto a_wiper = epee::misc_utils::create_scope_leave_handler([&]{
+    memwipe(static_cast<rct::key *>(a.data()), a.size() * sizeof(rct::key));
+  });
+  if (not reconstruction) {
+    a.resize(num_sources);
+    rv.p.pseudoOuts.resize(num_sources);
+    a[num_sources - 1] = rct::zero();
+    for (std::size_t i = 0; i < num_destinations; ++i) {
+      sc_add(
+        a[num_sources - 1].bytes,
+        a[num_sources - 1].bytes,
+        output_amount_masks[i].bytes
+      );
+    }
+    for (std::size_t i = 0; i < num_sources - 1; ++i) {
+      rct::skGen(a[i]);
+      sc_sub(
+        a[num_sources - 1].bytes,
+        a[num_sources - 1].bytes,
+        a[i].bytes
+      );
+      rct::genC(rv.p.pseudoOuts[i], a[i], sources[i].amount);
+    }
+    rct::genC(
+      rv.p.pseudoOuts[num_sources - 1],
+      a[num_sources - 1],
+      sources[num_sources - 1].amount
+    );
+  }
+  // check balance if reconstructing the tx
+  else {
+    rv.p.pseudoOuts = unsigned_tx.rct_signatures.p.pseudoOuts;
+    if (num_sources != rv.p.pseudoOuts.size())
+      return false;
+    rct::key balance_accumulator = rct::scalarmultH(rct::d2h(fee));
+    for (const auto& e: rv.outPk)
+      rct::addKeys(balance_accumulator, balance_accumulator, e.mask);
+    for (const auto& pseudoOut: rv.p.pseudoOuts)
+      rct::subKeys(balance_accumulator, balance_accumulator, pseudoOut);
+    if (not (balance_accumulator == rct::identity()))
+      return false;
+  }
+
+  // prepare input CLSAGs for signing
+  const rct::key message = get_pre_mlsag_hash(rv, hw::get_device("default"));
+
+  rv.p.CLSAGs.resize(num_sources);
+  if (reconstruction) {
+    if (num_sources != unsigned_tx.rct_signatures.p.CLSAGs.size())
+      return false;
+  }
+
+  cached_CLSAG.resize(num_sources);
+  if (not reconstruction)
+    cached_w.resize(num_sources);
+
+  for (std::size_t i = 0; i < num_sources; ++i) {
+    const std::size_t ring_size = rv.mixRing[i].size();
+    const rct::key& I = sources[i].multisig_kLRki.ki;
+    const std::size_t& l = sources[i].real_output;
+    if (l >= ring_size)
+      return false;
+    rct::keyV& s = rv.p.CLSAGs[i].s;
+    const rct::key& C_offset = rv.p.pseudoOuts[i];
+    rct::keyV P(ring_size);
+    rct::keyV C_nonzero(ring_size);
+
+    if (not reconstruction) {
+      s.resize(ring_size);
+      for (std::size_t j = 0; j < ring_size; ++j) {
+        if (j != l)
+          s[j] = rct::skGen();  //make fake responses
```

Comment:
At a glance, this doesn't seem correct. In the source paper (Musig2), the randomly point (`R`) is chosen cooperatively so that secret shares never leak to the other participants. It seems like something similar would have to be done here, but I haven't gone through the math details to verify any potential issue.

Has the adaption of musig2 been reviewed by anyone else (besides me)?

Reply:
This conditional is generating the fake CLSAG responses for all decoy ring members.

Reply:
I will probably finish up my review tomorrow, the hardest part is done (going through the relevant papers to track how it mapped to code, and changes necessary for the custom CLSAG scheme). So far it appears to be conforming to the two relevant papers; my review is "downhill" from here.

Reply:
I don't have anything else to add, but there are some comments that the OP should comment on or update.

Reply:
> but there are some comments that the OP should comment on or update.

Care to elaborate?

Reply:
* An aliasing issue that can be trivial removed
* The new varint function isn't needed (chop it for maintenance purposes)
* A `scalarmult` can be easily removed
* A `std::move` can be added easily
* A simple `const` can be added

They are all simple changes, so it seemed easiest to make them now. Only the first two are useful before merge. The varint thing in particular would need to be chopped before a release, otherwise it becomes part of the protocol and must be maintained. The simple swap+`memcpy` is preferable to the maintenance of that specific function.

Reply:
Oh, I thought you meant code comments.